src/transactions/BumpExpirationOpFrame.cpp

```diff
@@ -0,0 +1,167 @@
+// Copyright 2023 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+
+#ifdef ENABLE_NEXT_PROTOCOL_VERSION_UNSAFE_FOR_PRODUCTION
+#include "transactions/BumpExpirationOpFrame.h"
+
+namespace stellar
+{
+
+struct BumpExpirationMetrics
+{
+    medida::MetricsRegistry& mMetrics;
+
+    size_t mLedgerWriteByte{0};
+
+    BumpExpirationMetrics(medida::MetricsRegistry& metrics) : mMetrics(metrics)
+    {
+    }
+
+    ~BumpExpirationMetrics()
+    {
+        mMetrics
+            .NewMeter({"soroban", "bump-exp-op", "write-ledger-byte"}, "byte")
+            .Mark(mLedgerWriteByte);
+    }
+};
+
+BumpExpirationOpFrame::BumpExpirationOpFrame(Operation const& op,
+                                             OperationResult& res,
+                                             TransactionFrame& parentTx)
+    : OperationFrame(op, res, parentTx)
+    , mBumpExpirationOp(mOperation.body.bumpExpirationOp())
+{
+}
+
+bool
+BumpExpirationOpFrame::isOpSupported(LedgerHeader const& header) const
+{
+    return header.ledgerVersion >= 20;
+}
+
+bool
+BumpExpirationOpFrame::doApply(AbstractLedgerTxn& ltx)
+{
+    throw std::runtime_error("BumpExpirationOpFrame::doApply needs Config");
+}
+
+bool
+BumpExpirationOpFrame::doApply(Application& app, AbstractLedgerTxn& ltx,
+                               Hash const& sorobanBasePrngSeed)
+{
+    BumpExpirationMetrics metrics(app.getMetrics());
+
+    auto const& resources = mParentTx.sorobanResources();
+    auto const& footprint = resources.footprint;
+
+    UnorderedMap<LedgerKey, uint32_t> originalExpirations;
+
+    for (auto const& lk : footprint.readOnly)
+    {
+        // When we move to use EXPIRATION_EXTENSIONS, this should become a
+        // loadWithoutRecord, and the metrics should be updated.
+        auto ltxe = ltx.load(lk);
+        if (ltxe)
+        {
+            auto keySize = xdr::xdr_size(lk);
+            auto entrySize = xdr::xdr_size(ltxe.current());
+            metrics.mLedgerWriteByte += keySize;
+            metrics.mLedgerWriteByte += entrySize;
+
+            // Divide the limit by 2 for the metadata check since the previous
+            // state is also included in the meta.
+            if (resources.extendedMetaDataSizeBytes / 2 <
+                    metrics.mLedgerWriteByte ||
+                resources.readBytes < metrics.mLedgerWriteByte ||
+                resources.writeBytes < metrics.mLedgerWriteByte)
```

Comment:
This is suprising. If the footprint is RO, then only read bytes should be non-0. If we charge for writing the whole entries, it would make more sense if RW footprint was used instead.

Reply:
When we move to writing EXPIRATION_EXTENSIONS, we won't be writing these entries. EXPIRATION_EXTENSIONS will not be in the footprint though, so we'll have to figure out how to charge for those writes.

Reply:
By the testnet release.

Reply:
@dmkozh Should we be charging for instructions?

Reply:
Ideally we'd charge for something, but since metering logic is encapsulated in host now, I'm really not sure how much we should do in C++ (chances are it won't be accurate but also is cheap enough to not worry much).